File: ios/ReservationApp/ReservationApp/Data/Repositories/HomeRepository.swift
//
//  HomeRepository.swift
//  ReservationApp
//
//  Created by Mehmet Ali Kısacık on 13.11.2025.
//

import Foundation

class HomeRepository: HomeRepositoryProtocol {
    private let apiClient: APIClient
    
    nonisolated init(apiClient: APIClient = APIClient.shared) {
        self.apiClient = apiClient
    }
    
    func getHomeStats() async throws -> HomePageStats {
        return try await apiClient.request(
            endpoint: .homeStats,
            responseType: HomePageStats.self
        )
    }
    
    func getCategories() async throws -> [MenuCategory] {
        return try await apiClient.request(
            endpoint: .menuCategories,
            responseType: [MenuCategory].self
        )
    }
    
    func getMeals(restaurantId: String? = nil, categoryId: String? = nil) async throws -> [Meal] {
        var queryParams: [String: String] = [:]
        
        if let restaurantId = restaurantId {
            queryParams["restaurantId"] = restaurantId
        }
        
        if let categoryId = categoryId {
            queryParams["categoryId"] = categoryId
        }
        
        return try await apiClient.request(
            endpoint: .meals,
            queryParams: queryParams.isEmpty ? nil : queryParams,
            responseType: [Meal].self
        )
    }
    
    func getTodayMenu() async throws -> [Menu] {
        let today = Date()
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        let dateString = formatter.string(from: today)
        
        return try await apiClient.request(
            endpoint: .menus,
            queryParams: ["date": dateString],
            responseType: [Menu].self
        )
    }
}

File: android/domain/src/main/java/com/reservationapp/domain/repository/HomeRepository.kt
package com.reservationapp.domain.repository

import com.reservationapp.core.common.Result
import com.reservationapp.domain.model.HomePageStats
import com.reservationapp.domain.model.Meal
import com.reservationapp.domain.model.MealTimeSlot
import com.reservationapp.domain.model.Menu
import com.reservationapp.domain.model.MenuCategory
import com.reservationapp.domain.model.Restaurant

interface HomeRepository {
    suspend fun getHomeStats(): Result<HomePageStats>
    suspend fun getCategories(): Result<List<MenuCategory>>
    suspend fun getMeals(restaurantId: String? = null, categoryId: String? = null): Result<List<Meal>>
    suspend fun getTodayMenu(): Result<List<Menu>>
    suspend fun getRestaurants(): Result<List<Restaurant>>
    suspend fun getMealTimeSlots(): Result<List<MealTimeSlot>>
    suspend fun getMenus(date: String, restaurantId: String): Result<List<Menu>>
}

